platforms/software/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/ExcludingVariantArtifactSet.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resolve.resolver;
+
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvableArtifact;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedArtifactSet;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedVariant;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ExcludeSpec;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.component.model.VariantResolveMetadata;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A {@link ResolvedVariant} that applies artifact exclusions to a delegate {@link ResolvedVariant}.
+ */
+public class ExcludingVariantArtifactSet implements ResolvedVariant, VariantResolveMetadata.Identifier {
+
+    private final ResolvedVariant delegate;
+    private final ModuleIdentifier moduleId;
+    private final ExcludeSpec exclusions;
+
+    public ExcludingVariantArtifactSet(ResolvedVariant delegate, ModuleIdentifier moduleId, ExcludeSpec exclusions) {
+        this.delegate = delegate;
+        this.moduleId = moduleId;
+        this.exclusions = exclusions;
+    }
+
+    @Override
+    public DisplayName asDescribable() {
+        return Describables.of(delegate.asDescribable(), exclusions);
+    }
+
+    @Nullable
+    @Override
+    public VariantResolveMetadata.Identifier getIdentifier() {
+        // Since we use ourselves as the identifier, we must implement equals and hashCode
+        return this;
```

Comment:
🤔 I really dislike this pattern of classes implementing identifiers for themselves.  Generally, it's understood that identifiers are lightweight immutable data classes and are used with this understanding.  Making the type itself the identifier is counterintuitive and can result in nonsensical usage like `Map<Something.Identifier, Something>` which end up storing the same instance as both keys and values. 

It seems like it's the identifier which would better benefit from delegation.  An excluding identifier would contain an identifier, and an exclude spec and just delegate to the existing `equals/hashCode` on those objects and return the results.  This would keep it lightweight, and simplify this `ExcludingVariantArtifactSet` class further.

Reply:
Done